<h1><center>Insight into CO2 Emissions</center></h1>
<h4><center>
<a style="text-decoration:none" href="https://author1.github.io/">Harshil Purohit</a>, IIT Gandhinagar, <a style="text-decoration:none" href="mailto:harshil.p>@iitgn.ac.in">harshil.p@iitgn.ac.in</a>
<br><br>
<a style="text-decoration:none" href="https://author2.github.io/">Mihir Chauhan</a>, IIT Gandhinagar, <a style="text-decoration:none" href="mailto:mihir.hc@iitgn.ac.in">mihir.hc@iitgn.ac.in</a>
<br><br>
<a style="text-decoration:none" href="https://divyanshu0x16.github.io/">Divyanshu Meena</a>, IIT Gandhinagar, <a style="text-decoration:none" href="mailto:divyanshu.m@iitgn.ac.in">divyanshu.m@iitgn.ac.in</a>
</center></h4>

<center><a href="https://github.com/CS328-Spring-2022/writing-assignment-divyanshu_mihir_harshil"><img src="https://img.shields.io/badge/Github-Writing%20Assignment-brightgreen" alt="Repo"></a></center>

# CO2 Emissions Dataset

In [46]:
# Common imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import plotly.express as px

Adding mapping between country code and continents

In [47]:
iso_continents_df = pd.read_csv("./data/country-and-continent-codes-list.csv.gz", compression="gzip")
iso_continents_df.rename(columns = {'5ecdb312191a3559cde51177f4fb3b35.csv': 'Continent'}, inplace=True)
required_columns = ['Three_Letter_Country_Code', 'Continent']
iso_continents_df = iso_continents_df[required_columns]
iso_continents = dict(iso_continents_df.values)


In [48]:
co2_df = pd.read_csv("./data/owid-co2-data.csv.gz", compression="gzip")
co2_df.rename(columns = {'286b262ed4f75089c2a23c4ed4782e1f.csv': 'iso_code'}, inplace=True)
co2_df.sample(5)

iso_code        country    year       co2  co2_per_capita  trade_co2  \
2238        BRB       Barbados  1931.0     0.004           0.021        NaN   
24864  OWID_WRL          World  1954.0  6789.936           2.492        NaN   
16243       NaN  North America  1850.0    19.852           0.761        NaN   
21044       SSD    South Sudan  2010.0     1.313           0.138        NaN   
12282       KAZ     Kazakhstan  2004.0   188.547          12.364    -48.845   

       cement_co2  cement_co2_per_capita  coal_co2  coal_co2_per_capita  ...  \
2238          NaN                    NaN       NaN                  NaN  ...   
24864      97.083                  0.036  4039.060                1.482  ...   
16243         NaN                    NaN    19.852                0.761  ...   
21044         NaN                    NaN       NaN                  NaN  ...   
12282       0.346                  0.023   109.963                7.211  ...   

       ghg_excluding_lucf_per_capita  methane  methane_per_capita  \
2238                             NaN      NaN                 NaN   
24864                            NaN      NaN                 NaN   
16243                            NaN      NaN                 NaN   
21044                          7.216    48.61               5.112   
12282                         12.713    33.84               2.219   

       nitrous_oxide  nitrous_oxide_per_capita    population           gdp  \
2238             NaN                       NaN  1.760370e+05           NaN   
24864            NaN                       NaN  2.724847e+09           NaN   
16243            NaN                       NaN  3.975763e+07           NaN   
21044          21.07                     2.216  9.508372e+06           NaN   
12282           9.55                     0.626  1.525002e+07  2.177132e+11   

       primary_energy_consumption  energy_per_capita  energy_per_gdp  
2238                          NaN                NaN             NaN  
24864                         NaN                NaN             NaN  
16243                         NaN                NaN             NaN  
21044                         NaN                NaN             NaN  
12282                     501.997          32917.829           3.691  

[5 rows x 60 columns]

In [49]:
required_columns = ['iso_code', 'country', 'co2_per_capita', "year", "co2"]
globe_df = co2_df[required_columns]

globe_df = globe_df.dropna()
globe_df["year"] = globe_df["year"].astype("int")
# globe_df.sample(5)

Carbon Dioxide has been a significant cause of pollution and the primary driver of global climate change. It is commonly acknowledged that to avoid the effects of climate change; the world must reduce emissions as soon as possible. But, how this responsibility is shared between regions, countries, and individuals is a question we need to look into. 

The different metrics related to CO2 emissions, such as Net CO2 emissions, CO2 emissions per person, and CO2 emissions from various resources, each tell a different story. We also decided to look into the effects of economic policies and natural resource distribution on CO2 emissions.

Carbon dioxide emissions are measured in a million tonnes in all of the plots shown below. CO2 per unit of energy is measured in kilograms per kilowatt-hour. Carbon dioxide produced from specific sources such as coal, gas, cement, flaring, etc., is also measured in a million tonnes. CO2 per capita is measured in a million tonnes per person. 

# CO2 Emissions vs Time

In [50]:
globe_df['continent'] = globe_df['iso_code'].map(iso_continents)
globe_df_curr = globe_df[globe_df.year == 2019]
globe_df = globe_df.iloc[np.where(globe_df['continent'].notnull())[0]] #Remove values where continent is NaN
globe_df_curr = globe_df_curr.iloc[np.where(globe_df_curr['continent'].notnull())[0]]
globe_df = globe_df.sort_values('year')

In [51]:
fig = px.scatter_geo(globe_df_curr, locations="iso_code", color="continent",
                     hover_name="country", size="co2_per_capita",
                     projection="natural earth", template="plotly_dark")
fig.update_layout(title='CO2 Emissions Per Capita in 2019', legend_title='Continents')
fig.show()

In [52]:
fig = px.scatter_geo(globe_df_curr, locations="iso_code", color="continent",
                     hover_name="country", size="co2",
                     projection="natural earth", template="plotly_dark")
fig.update_layout(title='CO2 Emissions in 2019', legend_title='Continents')
fig.show()

In [53]:
time_df = co2_df[co2_df['country'].isin(['United Kingdom', 'United States', 'China', 'India', 'Russia', 'World'])]


time_df = time_df[['co2', 'year', 'country']]

#Add Europe
europe_data = co2_df[co2_df['country'] == 'Europe'][['co2', 'year', 'country']]
time_df = pd.concat([time_df, europe_data])

time_df = time_df.sort_values("year")
time_df = time_df[time_df['year'] > 1800]


fig = px.line(time_df, x='year', y = 'co2', color='country', title='CO2 Emissions vs Time')
fig.update_layout(xaxis_title="Year", template="plotly_dark",yaxis_title="CO2 Emissions", legend_title='Countries')
fig.show()

In [54]:
time_df = co2_df[co2_df['country'].isin(['United Kingdom', 'United States', 'China', 'India', 'Russia', 'World'])]


time_df = time_df[['co2_per_capita', 'year', 'country']]

#Add Europe
europe_data = co2_df[co2_df['country'] == 'Europe'][['co2_per_capita', 'year', 'country']]
time_df = pd.concat([time_df, europe_data])

time_df = time_df.sort_values("year")
time_df = time_df[time_df['year'] > 1800]


fig = px.line(time_df, x='year', y = 'co2_per_capita', color='country', title='CO2 Emissions  Per Capita vs Time')
fig.update_layout(xaxis_title="Year", template="plotly_dark",yaxis_title="CO2 Per Capita", legend_title='Countries')
fig.show()

### Impact of COVID-19
 
The Covid-19 pandemic and resulting economic crisis had an impact on global CO2.
The global CO2 emissions declined by 5.8% in 2020. It went from 36702.503 million tonnes to 34807.259 million tonnes. It was almost 2000 million tonnes of CO2, and the largest ever decline followed the global pandemic crisis. This decline is the equivalent of removing all of the European Union's CO2 emissions. Though China was one of the very few countries where emissions increased. The United States and Europe had an apparent decline in CO2 emissions.  

The 2009 economic crisis also had a similar impact on the global CO2 emission. It declined by almost 1.5% in 2009. Though the CO2 emissions quickly went up in 2010. There were around 500 million tonnes fewer emissions of CO2. We can see all major world events have impacted CO2 emissions, whether it is the end of the cold war or the end of the World War. We can spot different crisis effects on Global CO2 emissions. 

### Per capita CO2 emissions

The CO2 emission per capita is defined as the contribution in CO2 emission of an average citizen of each country. It is calculated by dividing the country’s total emissions by its population. In bellow visualization, we can see the differences in per capita emissions worldwide. 

The world’s largest per capita emissions CO2 emitters are the major oil-producing countries, as these countries mostly have relatively low populations. However, these countries have relatively low total annual emissions. More populated countries with high per capita emissions have high total yearly emissions, such as the United States, Canada, and Australia. Most of the European countries have per capita emissions near the world average. Many countries in Africa, South Asia, and South America still have very low per capita CO2. The lowest per capita emission is detected in Sub-Saharan African countries.

### Outsourcing CO2 Emissions

If we look at the plot of CO2 emissions vs. time for rich countries such as the United States or Europe, we notice that it has been decreasing steadily in the recent year. It looks like a good sign in the fight against global warming on a surface level. On the other hand, though, emissions in countries like China and India are increasing. This raises the question of whether that is this steady decline in rich countries and growth in developing countries related to each other. 

The answer is yes, as ‘rich’ countries indeed are ‘outsourcing’ their climate pollution to the developing countries by shifting their factories overseas. In the plot, we can notice a sharp increase in CO2 emissions of China since the year 2000. But a significant portion of these CO2 emissions is due to making goods for exporting. In other middle-income countries, such as India, we notice a more gradual rise in emissions.  

China's and other emerging economies' factories now emit more carbon pollution than the United States and Europe industries. But the per capita emissions in these economies are still below America and Europe. China and India regularly cite this gap in per capita emissions at climate talks to deflect the pressure to cut emissions. 

Outsourcing has made it challenging to hold countries accountable for their emissions. We can look at China and India and blame them, but at the same time, the pollution is caused due to goods that are being produced there for American demand. 

# Country-Wise Pie Chart for CO2 Emission in 2019

In [55]:
globe_df_2019 = globe_df[globe_df["year"] == 2019].sort_values("co2", ascending=False)
main_data = globe_df_2019.head(15).copy()

# Add all 16-rest countries as other
other = globe_df_2019.iloc[15:]["co2"]

# Append row containing other data
main_data.loc[len(main_data.index)] = ["OTH", "Other", 0, 2019, other.sum(), "Other"]
main_data

iso_code        country  co2_per_capita  year        co2      continent
4722       CHN          China           7.316  2019  10489.989           Asia
24053      USA  United States          15.972  2019   5255.816  North America
10976      IND          India           1.922  2019   2625.968           Asia
18989      RUS         Russia          11.513  2019   1679.449           Asia
12035      JPN          Japan           8.718  2019   1105.929           Asia
11294      IRN           Iran           8.845  2019    733.366           Asia
9589       DEU        Germany           8.518  2019    711.428         Europe
11108      IDN      Indonesia           2.441  2019    660.594           Asia
20982      KOR    South Korea          12.650  2019    648.025           Asia
19661      SAU   Saudi Arabia          18.163  2019    622.413           Asia
4279       CAN         Canada          15.567  2019    582.390  North America
3485       BRA         Brazil           2.302  2019    485.793  South America
20729      ZAF   South Africa           8.127  2019    475.923         Africa
14562      MEX         Mexico           3.414  2019    435.512  North America
1594       AUS      Australia          16.447  2019    414.517        Oceania
15         OTH          Other           0.000  2019   8502.505          Other

In [56]:
px.pie(main_data, names="country", values="co2", template="plotly_dark",title="Country-Wise CO2 Emission for 2019", hole=.3)

# CO2 Emission Breakup

In [57]:
req_cols = ['iso_code', 'country', "year", "co2", "trade_co2", "cement_co2", "coal_co2", "flaring_co2", "gas_co2", "oil_co2", "other_industry_co2"]

pie_breakup_df = co2_df[req_cols]
pie_breakup_df

pie_breakup_df = pie_breakup_df.dropna(subset=['iso_code', 'country', "year", "co2"])
pie_breakup_df["year"] = pie_breakup_df["year"].astype("int")

In [58]:
pie_breakup_ind_df = pie_breakup_df[(pie_breakup_df["iso_code"] == "IND") & (pie_breakup_df["year"] == 2019)][["trade_co2", "cement_co2", "coal_co2", "flaring_co2", "gas_co2", "oil_co2", "other_industry_co2"]].melt()

In [59]:
px.pie(pie_breakup_ind_df[pie_breakup_ind_df["value"] > 0], names="variable", values="value", template="plotly_dark",title="Breakup of CO2 Emission of India for 2019", hole=.3)

In [60]:
pie_breakup_usa_df = pie_breakup_df[(pie_breakup_df["iso_code"] == "USA") & (pie_breakup_df["year"] == 2019)][["trade_co2", "cement_co2", "coal_co2", "flaring_co2", "gas_co2", "oil_co2", "other_industry_co2"]].melt()

In [61]:
px.pie(pie_breakup_usa_df[pie_breakup_ind_df["value"] > 0], names="variable", values="value", template="plotly_dark",title="Breakup of CO2 Emission of USA for 2019", hole=.3)

In [62]:
pie_breakup_world = pie_breakup_df[(pie_breakup_df["year"] == 2019)][["trade_co2", "cement_co2", "coal_co2", "flaring_co2", "gas_co2", "oil_co2", "other_industry_co2"]].sum()
pie_breakup_world_df = pd.DataFrame({"variable": pie_breakup_world.index, "value": pie_breakup_world.values})

In [63]:
pie_breakup_world_df

variable      value
0           trade_co2     -4.110
1          cement_co2   3216.941
2            coal_co2  29144.329
3         flaring_co2    869.191
4             gas_co2  15106.597
5             oil_co2  23197.790
6  other_industry_co2    606.362

In [64]:
px.pie(pie_breakup_world_df[pie_breakup_world_df["value"] > 0], names="variable", values="value", template="plotly_dark",title="Breakup of CO2 Emission of World for 2019", hole=.3)

In [65]:
globe_df_2019.sort_values("co2_per_capita", ascending=False).head(20)

iso_code                          country  co2_per_capita  year  \
18564      QAT                            Qatar          40.619  2019   
15735      NCL                    New Caledonia          30.791  2019   
22723      TTO              Trinidad and Tobago          28.946  2019   
14855      MNG                         Mongolia          27.500  2019   
5494       CUW                          Curacao          23.014  2019   
3632       BRN                           Brunei          22.982  2019   
12516      KWT                           Kuwait          22.974  2019   
2158       BHR                          Bahrain          20.984  2019   
19661      SAU                     Saudi Arabia          18.163  2019   
20155      SXM        Sint Maarten (Dutch part)          16.620  2019   
1594       AUS                        Australia          16.447  2019   
24053      USA                    United States          15.972  2019   
12297      KAZ                       Kazakhstan          15.949  2019   
13626      LUX                       Luxembourg          15.838  2019   
23561      ARE             United Arab Emirates          15.781  2019   
4279       CAN                           Canada          15.567  2019   
8305       FRO                   Faeroe Islands          15.121  2019   
3180       BES  Bonaire Sint Eustatius and Saba          13.076  2019   
23175      TKM                     Turkmenistan          12.844  2019   
20982      KOR                      South Korea          12.650  2019   

            co2      continent  
18564   115.037           Asia  
15735     8.706        Oceania  
22723    40.378  North America  
14855    88.691           Asia  
5494      3.761  North America  
3632      9.958           Asia  
12516    96.651           Asia  
2158     34.439           Asia  
19661   622.413           Asia  
20155     0.705  North America  
1594    414.517        Oceania  
24053  5255.816  North America  
12297   295.869           Asia  
13626     9.752         Europe  
23561   154.186           Asia  
4279    582.390  North America  
8305      0.736         Europe  
3180      0.340  North America  
23175    76.319           Asia  
20982   648.025           Asia

### Country-Wise CO2 Emission

China emits the highest amount of carbon dioxide, around 30 % of the global carbon dioxide emission followed by the US and India contributing 15 % and 7% respectively. Combined, these three countries account for over half the global emissions. Even though China is a huge contributor to carbon dioxide emissions, its per capita emissions stand at a modest 7.3 tonnes/person which is half of the US per capita emissions which stand at 15.9 tonnes/person per year. 

The main sources of carbon emissions globally are coal, oil, and gas respectively. Emission due to the burning of coal is at a staggering 40%. 

In India, as coal is the main source of power generation, the disparity is even higher. Emission from coal accounts for 64% followed by oil which stands at 25%. However, if we consider the US which is the highest oil producing country we get a very different distribution which is expected. Emissions due to oil account for 44% whereas gas and coal contribute to 32% and 21% emissions respectively. 


# CO2 Per Unit Energy

In [66]:
time_df = co2_df[co2_df['country'].isin(['Europe (excl. EU-28)','Qatar','Africa','Iraq','United Arab Emirates','','United Kingdom', 'United States', 'China', 'India', 'Russia','World'])]


time_df = time_df[['co2_per_unit_energy', 'year', 'country']]



time_df = time_df.sort_values("year")
time_df = time_df[time_df['year'] > 1960]

#time_df
fig = px.line(time_df, x='year', y = 'co2_per_unit_energy', color='country', title='CO2 Emissions  Per unit energy vs Time')
fig.update_layout(xaxis_title="Year", template="plotly_dark",yaxis_title="CO2 Per unit energy", legend_title='Countries')
fig.show()

In [67]:
time_df = co2_df[co2_df['country'].isin(['Europe (excl. EU-28)','Qatar','Africa','Iraq','United Arab Emirates','','United Kingdom', 'United States', 'China', 'India', 'Russia','World'])]


time_df = time_df[['co2_per_unit_energy', 'year', 'country']]



time_df = time_df.sort_values("year")
time_df = time_df[time_df['year'] > 2000]

#time_df
fig = px.line(time_df, x='year', y = 'co2_per_unit_energy', color='country', title='CO2 Emissions  Per unit energy vs Time')
fig.update_layout(xaxis_title="Year", template="plotly_dark",yaxis_title="CO2 Per unit energy", legend_title='Countries')
fig.show()

CO2 emissions per unit of energy are annual production-based emissions of CO2, measured in kilograms per kilowatt-hour of primary production. It shows the efficient use of CO2, low being the most efficient system. With the advancement in technology, the CO2 emissions per unit of energy have shown a significant decrease, as we can see in the below visualization. We can also say that more advanced countries like the United States, the UK, and other countries, had significantly fewer emissions per energy than the UAE, Qatar, and Iraq. With more technological ideas shared over the world, most countries could reduce their emissions per energy. In recent times, most countries have had emissions per unit of energy between 0.1 to 0.4. India and Chain have been around 0.25 emissions per unit of energy, whereas UAE reduced it to about 0.1. 